src/mgmt/srv_target.c

```diff
@@ -657,3 +658,194 @@ ds_mgmt_hdlr_tgt_destroy(crt_rpc_t *td_req)
 	td_out->td_rc = rc;
 	crt_reply_send(td_req);
 }
+
+/*
+ * Return a URI string, like "ofi+sockets://192.168.1.70:44821", that callers
+ * are responsible for freeing with D_FREE.
+ */
+static int
+create_tag_uri(const char *base_uri, int i, char **uri)
+{
+	char   *type = "ofi+sockets:";
+	char   *u;
+	char   *p;
+	int	port;
+
+	/* TODO: Support other URI types. */
+	if (strncmp(base_uri, type, strlen(type)) != 0) {
+		D_ERROR("URI %s type not supported\n", base_uri);
+		return -DER_NOSYS;
+	}
+
+	/* Locate the ":" between the host and the port. */
+	p = strrchr(base_uri, ':');
+	if (p == NULL) {
+		D_ERROR("no ':' in base URI %s\n", base_uri);
+		return -DER_INVAL;
+	}
+
+	/* Calculate the port. */
+	port = atoi(p + 1) + i;
+	if (port > 65535 /* maximal port number */) {
+		D_ERROR("port %d out of range\n", port);
+		return -DER_INVAL;
+	}
+
+	/* Print the base URI with the port into u. */
+	D_ASPRINTF(u, "%.*s%d", (int)(p + 1 - base_uri), base_uri, port);
+	if (u == NULL)
+		return -DER_NOMEM;
+
+	*uri = u;
+	return 0;
+}
+
+static int
+add_server(crt_group_t *group, struct server_entry *server)
+{
+	int i;
+
+	D_DEBUG(DB_MGMT, "rank=%u uri=%s nctxs=%u\n", server->se_rank,
+		server->se_uri, server->se_nctxs);
+
+	for (i = 0; i < server->se_nctxs; i++) {
+		crt_node_info_t	info;
+		int		rc;
+
+		rc = create_tag_uri(server->se_uri, i, &info.uri);
+		if (rc != 0)
+			return rc;
+		rc = crt_group_node_add(group, server->se_rank, i, info);
+		D_FREE(info.uri);
+		if (rc != 0) {
+			D_ERROR("failed to add rank=%u tag=%d uri=%s: %d\n",
+				server->se_rank, i, info.uri, rc);
+			return rc;
+		}
+	}
```

Comment:
Do you need release the resource, for example uri, once there are failures?

Reply:
info.uri is freed after the crt_group_node_add call. But, errh, I used it later in the debug message; will fix.